training/replace_parameters.js

```diff
@@ -0,0 +1,395 @@
+                if (seq.some((w) => !isGoodWord(w))) {
+                    attempts -= 1;
+                    continue;
+                }
+                return seq.join(' ');
+            }
+
+            if (words.some((w) => !isGoodWord(w))) {
+                attempts -= 1;
+                continue;
+            }
+
+            if (words.length > 10) {
+                // if we keep getting samples that are too long, make one last ditch
+                // attempt by taking a long substring of the sample
+                if (attempts === 1) {
+                    const sampledLengthIdx = categorical([0.1, 0.3, 0.3, 0.3]);
+                    const length = [7, 8, 9, 10][sampledLengthIdx];
+                    const idx = Math.floor(this._rng() * (words.length-length));
+
+                    let seq = words.slice(idx, idx+length);
+                    if (seq.some((w) => !isGoodWord(w))) {
+                        attempts -= 1;
+                        continue;
+                    }
+                    return seq.join(' ');
+                } else {
+                    attempts -= 1;
+                    continue;
+                }
+            }
+            return sampled;
+        }
+        return null;
+    }
+
+    async _replaceTokensInSentence(id, sentence, parameters, replacements) {
+        let output = [];
+
+        for (let token of sentence) {
+            if (replacements.has(token)) {
+                output.push(replacements.get(token));
+            } else if (isReplaceToken(token)) {
+                if (!parameters.has(token)) {
+                    console.error(`error at ${id}: constant is in sentence but not in program`);
+                    output.push(token);
+                    continue;
+                }
+                const replace = await this._sampleParam(parameters.get(token));
+                if (!replace) {
+                    output.push(token);
+                } else {
+                    replacements.set(token, replace);
+                    output.push(replace);
+                }
+            } else {
+                output.push(token);
+            }
+        }
+        return output;
+    }
+
+    _replaceTokensInProgram(program, replacements) {
+        let output = [];
+        for (let token of program) {
+            if (replacements.has(token)) {
+                output.push('"', replacements.get(token), '"');
+                if (token.startsWith('HASHTAG_'))
+                    output.push('^^tt:hashtag');
+                else if (token.startsWith('USERNAME_'))
+                    output.push('^^tt:username');
+                else if (token.startsWith('GENERIC_ENTITY_'))
+                    output.push('^^' + token.substring('GENERIC_ENTITY_'.length, token.length-2));
+            } else {
+                output.push(token);
+            }
+        }
+        return output;
+    }
+
+    process(example) {
+        const sentence = example.preprocessed.split(' ');
```

Comment:
in case of line break or multiple space? `example.split(/(\s+)/).filter( (e) => e.trim().length > 0 );`